### ISSUE TITLE
Breakpad Crash uploaded as incorrect version when sent after updating the app

### ISSUE THREAD

**Issue by bobdoe132** [microsoft/appcenter-sdk-android#1265]
1. Crash the app
2. Update the app to newer version code
3. Crash uploaded as incorrect newer version code
---------------------------------------
1. Which SDK version are you using?
    com.microsoft.appcenter:appcenter-crashes:2.3.0
2. Which OS version did you experience the issue on?
    min 19, compile 29

**Comment by AnastasiaKubova** [comment]
Hi, @bobdoe132! Thank you for contacting us!
Could you please share the verbose [logs](https://docs.microsoft.com/en-us/appcenter/sdk/other-apis/android#adjust-the-log-level)?

**Comment by bobdoe132** [comment]
[appcenter.txt](https://github.com/microsoft/appcenter-sdk-android/files/3682661/appcenter.txt)

**Comment by bobdoe132** [comment]
The crash from build 8358 is uploaded as a crash from build 9000 which is wrong.

**Comment by AnastasiaKubova** [comment]
Hey @bobdoe132  ! Thank you for the provided logs. Unfortunately, I couldn't reproduce this issue. Could you please describe in detail steps for reproducing?

**Comment by bobdoe132** [comment]
See https://docs.microsoft.com/en-us/appcenter/sdk/crashes/android#reporting-ndk-crashes
On how we setup crash handling for the NDK.

0) Install / Run App Using "adb install APP.apk"
1) Crash App Using "std::terminate();"
2) Crash Is Now Saved But Hasn't Uploaded Yet
3) Update Version Name And Version Code Of App
4) Update App To New Version Using "adb install -r UPDATED_APP.apk"
5) Launch App Again, For The First Time Since Crash, After Updating
6) AppCenter Uploads Crash Saved In (2) As Crash From App Update In (4)

**Comment by bobdoe132** [comment]
I haven't tested java crashes. I mostly concerned with native crashes.

I should have mentioned that initially, sorry.

**Comment by guperrot** [comment]
Hi,  Java crashes are indeed working differently than native crashes.
For Java we save device properties at crash time, so the version is snapshotted and thus has the correct behavior.
For c/c++, breakpad produces a minidump that we are aware of only after restart, restart that can happen after the app update and thus the wrong version is sent.
We would need to have a timestamp <-> version mapping in a history like on our apple SDK but we haven't done that for Android yet.

**Comment by martinbonaciugome** [comment]
@guperrot why don't you guys just tell breakpad to store the dumps in a directory, where the name of the directory has the version/versioncode/etc...? that's what we did when we were using hockeyapp. A timestamp is not reliable as it could be wrong. Also if you are using a timestamp on iOS, does this get mapped on the client or after the upload? If it's not on the client, then the MSErrorReport contains inaccurate data. These are super critcal bugs. Please escalate

**Comment by martinbonaciugome** [comment]
@guperrot wait and how does the timestamp mapping work at all if you have multiple live builds?

**Comment by guperrot** [comment]
The timestamp are the launch times basically. If you have launch time at T1, then restart at T3, the crash happening at T2 will use the device info from T1. I am not aware of reliability issues with that approach that we also use for session tracking or token sending in auth.

What do you mean by multiple live builds?

**Comment by martinbonaciugome** [comment]
if you have builds 1 and 2 active at the same time, and if the criteria is know which build it crashed is by timestamp, then if you crash on build 1, upgrade to build 2, then relaunch, it will say it crashed on build 2

**Comment by martinbonaciugome** [comment]
and if you guys have other meta data to properly resolve it on your servers, will [MSCrashes lastSessionCrashReport].device.appVersion report the correct version in the client in the above scenario?

**Comment by guperrot** [comment]
Hey, I still don't understand what you mean by having 2 builds active at the same time, this is not possible on Android, either app v1 runs or v2 runs, not both at the same time.

On iOS if you crash on v1, then upgrade to v2, the crash will correctly say v1 because we use the device history. When v1 launches it saves t1 - d1 (device info1), then it crashes at t2, then it restarts at t3 - v2, t2 > t1 but t2 < t3 so we select d1 to associate the crash to.

This works also with lastSessionCrashReport.

**Comment by martinbonaciugome** [comment]
I mean active builds that we support, not on the same device. so what happens if user's time changes between the launches? on AppCenter it shows wrong times all the time. numerous reports say things like the crashed happened 1 day/1 week in the future.

**Comment by guperrot** [comment]
Because as I said in https://github.com/microsoft/appcenter-sdk-android/issues/1265#issuecomment-538072561 we have not done that work for Android yet, this is what this bug is about, to fix it for breakpad the same way iOS fixed the problem long time ago.

**Comment by martinbonaciugome** [comment]
Im just guessing you guys didn't fix it on iOS either, as timestamps are not accurate for this, as users time drifts all the time

**Comment by guperrot** [comment]
How frequent is device clock changed to be in the future? First time someone is reporting this. If that is true then it's more than just crash info that could be wrong in the SDK and the scope of that is more than just native crashes.

**Comment by guperrot** [comment]
It would impact everything, session, distribute state, authentication, data etc... so this discussion about timestamps should be discussed in a separate issue.

**Comment by martinbonaciugome** [comment]
well as I've said, we have seen numerous reports of weird times. I'm just trying to help you guys out with fixing this critical bug on android, as using timestamps is harder to implement, and in inaccurate that just storing the info in the directory path

**Comment by guperrot** [comment]
But we use timestamps in many other places so that fix that you mention would be very partial. And the device information also snapshots more fields than just versionCode and versionName. If the claim is that device clock in the future happens often, then we need a more generic solution to that specific problem.

**Comment by martinbonaciugome** [comment]
![image](https://user-images.githubusercontent.com/6127051/69458074-65228900-0d23-11ea-99d3-2ea1bf622e9b.png)

**Comment by guperrot** [comment]
Thanks for the report, however this is not the original issue. Timestamps in the future have many impacts on all modules and not just native crashes' version. I will forward this issue to the team to see what we can do about that. It would be best if a new issue was created for better tracking on our side.

**Comment by martinbonaciugome** [comment]
sure. thanks for the discussion and  quickness to respond

**Comment by guperrot** [comment]
Turns out iOS has to use timestamps as we can't control path or error ID before it occurs with the current state of PLCrashReporter.

After internal debate, we agreed to have a different solution on Android that will rely on path names like suggested in https://github.com/microsoft/appcenter-sdk-android/issues/1265/#issuecomment-557613018.

However we decided to use a variant where the folder is just a one time guid and storing the full device json file inside as osApiLevel/osVersion/osBuild/timezoneOffset/carrierCountry/locale can also change (and will probably store more files in the future inside the folder to improve sessionId/userId mappings for NDK).
